### PR TITLE
Adding cache block data width support to bsg_cache

### PR DESCRIPTION
**Pull request by Songchun-Li**
This PR aims at fixing [#372](https://github.com/bespoke-silicon-group/basejump_stl/issues/372) 

The change is minimal and merely about block offset of address.

### CODE REVIEW

File: bsg_cache/bsg_cache.v [hunk 1]
```diff
@@ -70,7 +70,7 @@ module bsg_cache
   localparam lg_sets_lp=`BSG_SAFE_CLOG2(sets_p);
   localparam data_mask_width_lp=(data_width_p>>3);
   localparam lg_data_mask_width_lp=`BSG_SAFE_CLOG2(data_mask_width_lp);
-  localparam lg_block_size_in_words_lp=`BSG_SAFE_CLOG2(block_size_in_words_p);
+  localparam lg_block_size_in_words_lp= (block_size_in_words_p > 1) ? `BSG_SAFE_CLOG2(block_size_in_words_p) : 0;
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:73]: can just use $clog2

File: bsg_cache/bsg_cache_miss.v [hunk 2]
```diff
@@ -387,7 +390,7 @@ module bsg_cache_miss
         dma_addr_o = {
           addr_tag_v,
           addr_index_v,
-          addr_block_offset_v, // used for snoop data in dma.
+          {(block_size_in_words_p > 1){addr_block_offset_v}}, // used for snoop data in dma.
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache_miss.v:393]: only this line is necessary, revert others

File: bsg_cache/bsg_cache.v [hunk 3]
```diff
@@ -109,8 +108,12 @@ module bsg_cache
     = cache_pkt.addr[lg_data_mask_width_lp+lg_block_size_in_words_lp+lg_sets_lp+:lg_ways_lp];
   assign addr_index
     = cache_pkt.addr[lg_data_mask_width_lp+lg_block_size_in_words_lp+:lg_sets_lp];
-  assign addr_block_offset
-    = cache_pkt.addr[lg_data_mask_width_lp+:lg_block_size_in_words_lp];
+  if (block_size_in_words_p > 1)
+    begin
+      logic [lg_block_size_in_words_lp-1:0] addr_block_offset;
+      assign addr_block_offset
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:114]: This variable seems to be unused?

File: bsg_cache/bsg_cache.v [hunk 4]
```diff
@@ -160,13 +163,16 @@ module bsg_cache
...
 
   assign addr_index_tl =
     addr_tl_r[lg_data_mask_width_lp+lg_block_size_in_words_lp+:lg_sets_lp];
 
-  assign addr_block_offset_tl =
-    addr_tl_r[lg_data_mask_width_lp+:lg_block_size_in_words_lp];
+  if (block_size_in_words_p > 1)
+    begin
+      logic [lg_block_size_in_words_lp-1:0] addr_block_offset_tl;
+      assign addr_block_offset_tl =
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:173]: This variable seems to be unused?

File: bsg_cache/bsg_cache.v [hunk 5]
```diff
@@ -70,7 +70,7 @@ module bsg_cache
   localparam lg_sets_lp=`BSG_SAFE_CLOG2(sets_p);
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:70]: Changes to this file don't seem to have any functional impact?

File: bsg_cache/bsg_cache_miss.v [hunk 6]
```diff
@@ -157,8 +157,11 @@ module bsg_cache_miss
     = addr_v_i[lg_data_mask_width_lp+lg_block_size_in_words_lp+lg_sets_lp+:tag_width_lp];
   assign addr_way_v
     = addr_v_i[lg_sets_lp+lg_block_size_in_words_lp+lg_data_mask_width_lp+:lg_ways_lp];
-  assign addr_block_offset_v
-    = addr_v_i[lg_data_mask_width_lp+:lg_block_size_in_words_lp];
+  if (block_size_in_words_p > 1)
```
Comment (Songchun-Li) [inline bsg_cache/bsg_cache_miss.v:160]: > only this line is necessary, revert others

I think the if generate statement here is also necessary, as lg_block_size_in_words_lp can be 0 when block_size_in_words_p == 1.

File: bsg_cache/bsg_cache_miss.v [hunk 7]
```diff
@@ -17,7 +17,7 @@ module bsg_cache_miss
     ,parameter sets_p="inv"
     ,parameter ways_p="inv"
 
-    ,parameter lg_block_size_in_words_lp=`BSG_SAFE_CLOG2(block_size_in_words_p)
+    ,parameter lg_block_size_in_words_lp= (block_size_in_words_p > 1) ? `BSG_SAFE_CLOG2(block_size_in_words_p) : 0
```
Comment (Songchun-Li) [inline bsg_cache/bsg_cache_miss.v:20]: > only this line is necessary, revert others

I think it should be kept, otherwise the issues in bit slicing is still, like tag_width_lp is less than it should be.

File: bsg_cache/bsg_cache.v (hunk 5 above)
Comment (Songchun-Li) [reply bsg_cache/bsg_cache.v:70]: Yep, the main issue was lg_block_size_in_words_lp took up an additional 1 bit in bit slicing. The changes should not have any functional  impacts.

File: bsg_cache/bsg_cache.v (hunk 3 above)
Comment (Songchun-Li) [reply bsg_cache/bsg_cache.v:114]: addr_block_offset and addr_block_offset_tl are aunused. Shall I remove them?

File: bsg_cache/bsg_cache.v (hunk 5 above)
Comment (dpetrisko) [reply bsg_cache/bsg_cache.v:70]: Revert the changes if they have no impact

File: bsg_cache/bsg_cache_miss.v (hunk 7 above)
Comment (dpetrisko) [reply bsg_cache/bsg_cache_miss.v:20]: $clog2 rather than safeclog2, since we want the zero behavior

File: bsg_cache/bsg_cache.v (hunk 1 above)
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:73]: $clog2

File: bsg_cache/bsg_cache_dma.v [hunk 8]
```diff
@@ -18,7 +18,7 @@ module bsg_cache_dma
     ,parameter ways_p="inv"
     ,parameter dma_data_width_p=data_width_p
 
-    ,parameter lg_block_size_in_words_lp=`BSG_SAFE_CLOG2(block_size_in_words_p)
+    ,parameter lg_block_size_in_words_lp=(block_size_in_words_p > 1) ? `BSG_SAFE_CLOG2(block_size_in_words_p) : 0
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache_dma.v:21]: $clog2

File: bsg_cache/bsg_cache.v (hunk 3 above)
Comment (dpetrisko) [reply bsg_cache/bsg_cache.v:114]: These should be used below in the recover address, with the zero_width conditional

File: bsg_cache/bsg_cache.v (hunk 5 above)
Comment (dpetrisko) [reply bsg_cache/bsg_cache.v:70]: Make a variable block_offset_width_lp=(block_size_in_words_p > 1) ?lg_data_mask_width_lp+lg_block_size_in_words_lp : lg_data_mask_width_lp

And then use this variable throughout.  That will get rid of nearly all of the conditionals

File: bsg_cache/bsg_cache.v [hunk 9]
```diff
@@ -70,8 +70,9 @@ module bsg_cache
...
   localparam data_mask_width_lp=(data_width_p>>3);
   localparam lg_data_mask_width_lp=`BSG_SAFE_CLOG2(data_mask_width_lp);
-  localparam lg_block_size_in_words_lp=`BSG_SAFE_CLOG2(block_size_in_words_p);
-  localparam tag_width_lp=(addr_width_p-lg_data_mask_width_lp-lg_sets_lp-lg_block_size_in_words_lp);
+  localparam lg_block_size_in_words_lp=(block_size_in_words_p > 1) ? $clog2(block_size_in_words_p) : 0;
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:73]: revert this change and remove all conditionals below

File: bsg_cache/bsg_cache_dma.v [hunk 10]
```diff
@@ -18,7 +18,7 @@ module bsg_cache_dma
     ,parameter ways_p="inv"
     ,parameter dma_data_width_p=data_width_p
 
-    ,parameter lg_block_size_in_words_lp=`BSG_SAFE_CLOG2(block_size_in_words_p)
+    ,parameter lg_block_size_in_words_lp=(block_size_in_words_p > 1) ? $clog2(block_size_in_words_p) : 0
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache_dma.v:21]: Use SAFE_CLOG2 and block_offset_width_lp as in bsg_cache

File: bsg_cache/bsg_cache.v [hunk 11]
```diff
@@ -70,8 +70,9 @@ module bsg_cache
...
+  localparam block_offset_width_lp=lg_data_mask_width_lp+lg_block_size_in_words_lp;
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache.v:74]: ,parameter block_offset_width_lp=(block_size_in_words_p > 1) ? lg_data_mask_width_lp+lg_block_size_in_words_lp : lg_data_mask_width_lp

File: bsg_cache/bsg_cache_miss.v [hunk 12]
```diff
@@ -387,8 +388,7 @@ module bsg_cache_miss
         dma_addr_o = {
           addr_tag_v,
           addr_index_v,
-          addr_block_offset_v, // used for snoop data in dma.
-          {(lg_data_mask_width_lp){1'b0}}
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache_miss.v:391]: Don't change the behavior here. Use a zero_width signal

File: bsg_cache/bsg_cache_miss.v [hunk 13]
```diff
@@ -149,16 +150,16 @@ module bsg_cache_miss
...
+    = addr_v_i[block_offset_width_lp+:lg_sets_lp];
   assign addr_tag_v
-    = addr_v_i[lg_data_mask_width_lp+lg_block_size_in_words_lp+lg_sets_lp+:tag_width_lp];
+    = addr_v_i[block_offset_width_lp+lg_sets_lp+:tag_width_lp];
   assign addr_way_v
-    = addr_v_i[lg_sets_lp+lg_block_size_in_words_lp+lg_data_mask_width_lp+:lg_ways_lp];
+    = addr_v_i[block_offset_width_lp+lg_sets_lp+:lg_ways_lp];
   assign addr_block_offset_v
-    = addr_v_i[lg_data_mask_width_lp+:lg_block_size_in_words_lp];
+    = addr_v_i[0+:block_offset_width_lp];
```
Comment (dpetrisko) [inline bsg_cache/bsg_cache_miss.v:162]: don't change the behavior here, use a zero-width signal

**Review by dpetrisko — APPROVED** [review]
LGTM! Have you tested this configuration in the regression_v2 testing directory?

@tommydcjung , look good to you?

**Comment by Songchun-Li** [comment]
Yep, I ran thourgh the tests in both regression_64 and regression_v2.

**Review by taylor-bsg — COMMENTED** [review]
Good changes! I think what is missing is something added to the regression to test explicitly this case. @tommydcjung @dpetrisko Do you guys have a recommended way to do this? Should we have a parameter sweep over block size (and other parameters like associativity?) that is added to the existing regression?

**Comment by dpetrisko** [comment]
The testbenches are parameterized pretty well, so I think the cross product methodology in the other basejump regressions would work here.

That said, as long as the tests pass when manually changing the parameters, I think setting up a regression sweep is a different PR

**Comment by tommydcjung** [comment]
I think it's good, but you have to add regression for new features. There is already a facility to sweep.
https://github.com/bespoke-silicon-group/basejump_stl/blob/master/testing/bsg_cache/regression_v2/sweep.sh
I think you can add an additional parameter for block_size_in_words_p?
It would also be nice to think about how you can break this special case, and write some tests for it.

**Comment by dpetrisko** [comment]
>https://github.com/bespoke-silicon-group/basejump_stl/blob/master/testing/bsg_cache/regression_v2/sweep.sh

Nice!

**Comment by Songchun-Li** [comment]
When I tried to add additional test coverage over block_size_in_words_p == 1 and over data_width_p > 64 for #377, I found that there is some code in both testing modules and python scripts preventing me from changing the parameters in the v2 test, because it is written explicitly for the configuration of data_width_p == 32, block_size_in_words_p == 8.

For example, in the [basic_checker.v](https://github.com/bespoke-silicon-group/basejump_stl/blob/959aedfe1414453f5578419cb4019bee3873387b/testing/bsg_cache/regression_v2/basic_checker.v#L50), both the store mask and the mask for result data selection are fixed for data_width==32. Also, the [get_addr](https://github.com/bespoke-silicon-group/basejump_stl/blob/959aedfe1414453f5578419cb4019bee3873387b/testing/bsg_cache/regression_v2/test_base.py#L157) in the test_base.py and some other trace scripts have the same issue.

And my proposal is to parameterize block_size_in_words_p and data_width_p in both the basic checker module and the python scripts and make the v2 test support following configuration.
```
cache_block_width_p  |  data_width_p   |  block_size_in_words_p
                            32                 8
      256                   64                 4
                           128                 2
                           256                 1
```

@tommydcjung @dpetrisko Should I add this change to the PR for testing?

**Comment by dpetrisko** [comment]
It's @tommydcjung's call, but adding regression_128 in the vein of regression_64 might be a good idea, since 128-bit is a configuration we intend to support in BlackParrot. And then we would leave regression_v2 as a more intensive 32-bit regression

**Comment by Songchun-Li** [comment]
Agree, adding this change to regression_64 leads to less change in the code, and the cache block size is the same as the blackparrot one.

**Comment by tommydcjung** [comment]
More often replicating the bug in a small test turns out to be greater effort than the fixing the bug. But it's a worthwhile effort for many good reasons. I think it would be great to have regression testing infrastructure parameterizable, so that we can create more tests for specific configurations. Then, we don't need to create a entirely new testbench  that needs to be maintained each time we add a new configuration.

**Comment by dpetrisko** [comment]
@tommydcjung Sounds reasonable. I'm not too familiar with which testbenches do what.  Is the idea that we would combine regression, regression_64 and regression_v2 into a single parameterizable testbench?

How much work do you estimate that is? The alternative being adding the 128 bit testbench for #374 and #377 and cleaning up the testbenches in a future PR

**Comment by Songchun-Li** [comment]
Hi @tommydcjung,  @dpetrisko here is a summary of recent commits

- Changed the Makefile, test_base.py, testbench.v and basic_check.v to work with varying block_size_in_words_p and data_width_p. And all supported configurations are shown in the following table

```
cache_block_width_p  |  data_width_p   |  block_size_in_words_p
                            64                 8
      512                  128                 4
                           256                 2
                           512                 1
```

- Fixed the zero-width signal bug in bsg_nonsynth_dma_model when block_size_in_words_p==1
- Added a new test test_mask1 , to have an intensive test on LM/SM which are frequently uesd in black parrot  when data_width>64. 
- Added a regression scripts, width_sweep.sh, to sweep all configurations mentioned in the above table.

BTW, by testing the bsg_cache with the updated testbench, I think I spot another bug similar to the one in #377. The detailed description are added [here](https://github.com/bespoke-silicon-group/basejump_stl/pull/377#issuecomment-765884011).

File: testing/bsg_cache/regression_64/testbench.v [hunk 14]
```diff
@@ -143,7 +143,7 @@ module testbench();
 
 
   // trace replay
-  localparam rom_addr_width_lp = 26;
+  localparam rom_addr_width_lp = 23;
```
Comment (Songchun-Li) [inline testing/bsg_cache/regression_64/testbench.v:146]: Shrink the addr width so that the size of trace rom does not exceed the limitation of 2GB-1, while all the traces still fit in.

File: testing/bsg_cache/regression_64/testbench.v [hunk 15]
```diff
@@ -23,11 +23,11 @@ module testbench();
...
   // parameters
   localparam addr_width_p = 30;
-  localparam data_width_p = 64;
-  localparam block_size_in_words_p = 8;
+  localparam data_width_p = 512/`BLOCK_SIZE_IN_WORDS_P;
+  localparam block_size_in_words_p = `BLOCK_SIZE_IN_WORDS_P;
   localparam sets_p = 64;
   localparam ways_p = 8;
-  localparam mem_size_p = block_size_in_words_p*sets_p*ways_p*4;
+  localparam mem_size_p = 8*sets_p*ways_p*4;
```
Comment (Songchun-Li) [inline testing/bsg_cache/regression_64/testbench.v:30]: It is possible that the shadowmem in basic checker runs out of space when block_size_in_words_p<8, so I just keep it as 8 as an temporary solution here.

**Comment by tommydcjung** [comment]
Awesome changes and good bugs found.  I tried running the updated regression_64, and got these messages. Am I missing something?
```
start test
################
8
out/test_random1/simv.log:[BSG_FINISH] Test Successful.
out/test_random2/simv.log:[BSG_FINISH] Test Successful.
out/test_random3/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic1/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic2/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic3/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic4/simv.log:[BSG_FINISH] Test Successful.
out/test_stride1/simv.log:[BSG_FINISH] Test Successful.
out/test_mask1/simv.log:[BSG_FINISH] Test Successful.
################
4
out/test_random1/simv.log:[BSG_FINISH] Test Successful.
out/test_random2/simv.log:[BSG_FATAL] output does not match expected result. Id=        662, Expected: 0000000000000000000000000000002f. Actual: 000000000000002f000000000000002f.
out/test_random3/simv.log:[BSG_FATAL] output does not match expected result. Id=        543, Expected: 0000000000000000000000000000000b. Actual: 000000000000000b000000000000000b.
out/test_atomic1/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic2/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic3/simv.log:[BSG_FATAL] output does not match expected result. Id=        521, Expected: 00000000000000000000000000000005. Actual: 00000000000000050000000000000005.
out/test_atomic4/simv.log:[BSG_FATAL] output does not match expected result. Id=        534, Expected: 00000000000000000000000000000001. Actual: 00000000000000010000000000000001.
out/test_stride1/simv.log:[BSG_FATAL] output does not match expected result. Id=     445847, Expected: 00000000000000000000000000037fc6. Actual: 0000000000037fc60000000000037fc6.
out/test_mask1/simv.log:[BSG_FINISH] Test Successful.
################
2
out/test_random1/simv.log:[BSG_FINISH] Test Successful.
out/test_random2/simv.log:[BSG_FATAL] output does not match expected result. Id=        563, Expected: 0000000000000000000000000000000000000000000000000000000000000000. Actual: 0000000000000000000000000000000000000000000000100000000000000000.
out/test_random3/simv.log:[BSG_FATAL] output does not match expected result. Id=        520, Expected: 0000000000000000000000000000000000000000000000000000000000000002. Actual: 0000000000000002000000000000000200000000000000020000000000000002.
out/test_atomic1/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic2/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic3/simv.log:[BSG_FATAL] output does not match expected result. Id=        517, Expected: 0000000000000000000000000000000000000000000000000000000000000002. Actual: 0000000000000002000000000000000200000000000000020000000000000002.
out/test_atomic4/simv.log:[BSG_FATAL] output does not match expected result. Id=        516, Expected: 0000000000000000000000000000000000000000000000000000000000000003. Actual: 0000000000000003000000000000000300000000000000030000000000000003.
out/test_stride1/simv.log:[BSG_FATAL] output does not match expected result. Id=     445847, Expected: 0000000000000000000000000000000000000000000000000000000000037fc6. Actual: 0000000000037fc60000000000037fc60000000000037fc60000000000037fc6.
out/test_mask1/simv.log:[BSG_FINISH] Test Successful.
################
1
out/test_random1/simv.log:[BSG_FINISH] Test Successful.
out/test_random2/simv.log:[BSG_FATAL] output does not match expected result. Id=        596, Expected: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000. Actual: 00000000000000050000000000000005000000000000000500000000000000050000000000000005000000000000000500000000000000050000000000000005.
out/test_random3/simv.log:[BSG_FATAL] output does not match expected result. Id=        538, Expected: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000011. Actual: 00000000000000110000000000000011000000000000001100000000000000110000000000000011000000000000001100000000000000110000000000000011.
out/test_atomic1/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic2/simv.log:[BSG_FINISH] Test Successful.
out/test_atomic3/simv.log:[BSG_FATAL] output does not match expected result. Id=        526, Expected: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003. Actual: 00000000000000030000000000000003000000000000000300000000000000030000000000000003000000000000000300000000000000030000000000000003.
out/test_atomic4/simv.log:[BSG_FATAL] output does not match expected result. Id=        519, Expected: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004. Actual: 00000000000000040000000000000004000000000000000400000000000000040000000000000004000000000000000400000000000000040000000000000004.
out/test_stride1/simv.log:[BSG_FATAL] output does not match expected result. Id=     445847, Expected: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000037fc6. Actual: 0000000000037fc60000000000037fc60000000000037fc60000000000037fc60000000000037fc60000000000037fc60000000000037fc60000000000037fc6.
out/test_mask1/simv.log:[BSG_FINISH] Test Successful.

```

**Comment by tommydcjung** [comment]
Could you document a README.md in regression_64/ describing what are the commands people supposed to run to run the regression?

**Comment by Songchun-Li** [comment]
Because when block_size_in_words_p < 8, data_width_p > 64, therefore, the bugs for D ops in #377 leads to test failure, you will need the changes in that PR to pass the tests. Some tests are always successful because they don't have D size ops. 

In other word,  this testbench also meet the new test requirement for #377   

Sure, I will add a README soon.

**Comment by tommydcjung** [comment]
I would suggest adding a test (test_stride2) where you store M words in (2^x)-1 strides (e.g. 1,3,7,15, ...), and load back to test. Something like
```
for x in range(13):
  stride = (2^x)-1
  for i in range(2000):
    store(stride*i)
  for i in range(2000):
    load(stride*i)
```

**Comment by tommydcjung** [comment]
> Because when block_size_in_words_p < 8, data_width_p > 64, therefore, the bugs for D ops in #377 leads to test failure, you will need the changes in that PR to pass the tests. Some tests are always successful because they don't have D size ops.
> 
> In other word, this testbench also meet the new test requirement for #377
> 
> Sure, I will add a README soon.

Okay so are these two PRs going together? Maybe we can merge them together, since they are related problems?

**Comment by Songchun-Li** [comment]
Hi @tommydcjung, a README and the new test, test_stride2 are added. And I made some changes to test_stride, because ((2^x)-1)*2000 can execced the max address defined in test_base.py, but the major idea is the same.

File: testing/bsg_cache/regression_64/testbench.v [hunk 16]
```diff
@@ -23,11 +23,11 @@ module testbench();
...
   // parameters
   localparam addr_width_p = 30;
-  localparam data_width_p = 64;
-  localparam block_size_in_words_p = 8;
+  localparam data_width_p = 512/`BLOCK_SIZE_IN_WORDS_P;
+  localparam block_size_in_words_p = `BLOCK_SIZE_IN_WORDS_P;
   localparam sets_p = 64;
   localparam ways_p = 8;
-  localparam mem_size_p = block_size_in_words_p*sets_p*ways_p*4;
+  localparam mem_size_p = 2**(17-`BSG_SAFE_CLOG2(data_width_p/8));
```
Comment (Songchun-Li) [inline testing/bsg_cache/regression_64/testbench.v:30]: The max addr set in test_base.py is 2**17, so I make the mem_size_p consistent with it.

**Comment by Songchun-Li** [comment]
The bugs fixed in #377 are relevant to the changes in this PR, because data_width_p is larger than 64 when block_size_in_words_p == 1. Therefore, I got them merged by rebaseing the changes onto the ones in #377. This PR should be able to pass width regression test now.

**Comment by tommydcjung** [comment]
okay cool.
```
Lint-[CAWM-L] Width mismatch
/mnt/users/ssd1/homes/dcjung/bsg/basejump_stl/bsg_cache/bsg_cache.v, 624
  Continuous assignment width mismatch
  64 bits (lhs) versus 256 bits (rhs).
  Source info: assign atomic_64.amo32_mem_in = (ld_data_final_li[2] << 32);


Lint-[CAWM-L] Width mismatch
/mnt/users/ssd1/homes/dcjung/bsg/basejump_stl/bsg_cache/bsg_cache.v, 625
  Continuous assignment width mismatch
  64 bits (lhs) versus 256 bits (rhs).
  Source info: assign atomic_64.amo64_mem_in = ld_data_final_li[3];
```
could you also take care of this lint warnings?

**Review by tommydcjung — APPROVED** [review]

**Comment by Songchun-Li** [comment]
> could you also take care of this lint warnings?

The port width mismatch is resolved

**Comment by dpetrisko** [comment]
awesome work @Songchun-Li ! Thanks for the reviews @tommydcjung !
